corehq/apps/users/management/commands/migrate_user_history_to_new_structure.py

```diff
@@ -0,0 +1,734 @@
+
+# f"Primary location: {location.name}[{location.location_id}]"
+WEB_USER_PRIMARY_LOCATION_REGEX = re.compile(r'Primary location: (.+)\[(.+)]')
+
+# f"Assigned locations: {location_names}" where location names is a python list object of names
+COMMCARE_USER_ASSIGNED_LOCATIONS_REGEX = re.compile(r'Assigned locations: \[(.+)]')
+
+# f"Assigned locations: {locations_info}"
+# where locations_info is a ", " joined location_name[location_id] for locations
+WEB_USER_ASSIGNED_LOCATIONS_REGEX = re.compile(r'Assigned locations: (.*)')
+
+# f"Groups: {groups_info}" same as assigned locations
+GROUPS_REGEX = re.compile(r'Groups: (.*)')
+
+# f"Added as web user to domain '{domain}'"
+ADDED_AS_WEB_USER_REGEX = re.compile(r"Added as web user to domain '(\S+)'")
+
+# f"Invited to domain '{domain}'"
+INVITED_TO_DOMAIN_REGEX = re.compile(r"Invited to domain '(.+)'")
+
+# f"Invitation revoked for domain '{domain}'"
+INVITATION_REVOKED_FOR_DOMAIN_REGEX = re.compile(r"Invitation revoked for domain '(.+)'")
+
+# f'Two factor disabled. Verified by: {verified_by}, verification mode: "{verification_mode}"'
+DISABLED_WITH_VERIFICATION_REGEX = re.compile(
+    r'Two factor disabled. Verified by: (.+), verification mode: "(.+)"')
+
+# location-name[location_id]
+# group-name[group_id]
+NAME_WITH_ID_REGEX = re.compile(r'(.+)\[(.+)]')
+
+
+class Command(BaseCommand):
+    help = "Migrate User History records to new structure"
+
+    def add_arguments(self, parser):
+        parser.add_argument(
+            '--save',
+            action='store_true',
+            dest='save',
+            default=False,
+            help="actually save records else just log",
+        )
+        parser.add_argument(
+            '--skip-assertions',
+            action='store_true',
+            dest='skip_assertions',
+            default=False,
+            help="to be used to skip assertions in change message creation, useful for manual updates",
+        )
+        parser.add_argument(
+            '--only-pending',
+            action='store_true',
+            dest='only_pending',
+            default=False,
+            help="to iterate only logs that are pending to be migrated"
+        )
+
+    def handle(self, **options):
+        save = options['save']
+        skip_assertions = options['skip_assertions']
+        only_pending = options['only_pending']
+        if save and skip_assertions:
+            raise CommandError("It is not recommended to have save with assertions skipped")
+        wb = Workbook()
+        ws = wb.active
+        ws.append([
+            'ID', 'Message', 'Change Messages', 'Details Changes', 'Changes', 'Changes equal?',
+            'Details Changed Via', 'Changed Via']
+        )
+
+        records = UserHistory.objects
+        if only_pending:
+            # filter on the new column that is expected to be blank for not migrated records only
+            records = records.filter(changed_via='')
+
+        for user_history in records.all():
```

Comment:
Might want to paginate this queryset and the other - this'll pull all records into memory at once

Reply:
https://github.com/dimagi/commcare-hq/commit/452498d5477e6676cffd0ed407da9faf4f20ad7c